GetJobber/atlantis
# feat(components): Implement Custom Activator to Combobox

## Motivations


We need to allow consumers of `Combobox` to build their own custom activator, or be given a default `ComboboxTrigger`

## Changes



### Added

 
 Added a `ComboboxActivator` component which allows consumers to set the trigger of the `Combobox` to be a Button or Chip and it will assign the role of `combobox` to either trigger type.
  
 This way, consumers can customize the appearance of their trigger, as well as handle the output UI as they want to (in a form, separate chips in a card, etc).  If anything other than a `Chip` or `Button` is a child of the `ComboboxActivator`, it will not be given the role of 'combobox'.
 
 Implementation:
 ```
 <Combobox>
      <Combobox.Activator>
        <Button
          label="Select Teammate"
          variation="subtle"
          type="primary"
          icon="arrowDown"
          iconOnRight={true}
        />
      </Combobox.Activator>
      <Combobox.Content
        options={[
          { id: "1", label: "Bilbo Baggins" },
          ...rest of the code
```
 
 -----
Also added a `ComboboxTrigger`.  
When `ComboboxActivator` is not used, the `ComboboxTrigger` will be the default.

```
 <Combobox multiSelect heading="Tags">
      <Combobox.Content
        options={[
          { id: "1", label: "Bilbo Baggins" },
             ...rest of the code
```

`ComboboxTrigger` will:
- render as a Chip with 'add' suffix  
- will also display the selectedOption(s) as the Chip label as they are selected
- when a selection is made, the Chip variation will change from `subtle` to `base`

Once a selection is made in single select, the selected option will become the Chip label.
Once a selection is made in multi select, the Chip will keep the `heading` visible and the `label` will be the list of selectedOptions separated by a comma

Examples of `ComboboxTrigger`:
 
 #### Single select
 
 ![Kapture 2023-10-26 at 11 40 31](https://github.com/GetJobber/atlantis/assets/104797704/4c6d9dab-ebf3-4075-b112-5039aef3d57f)
 
 #### Multi select
 
![Kapture 2023-10-26 at 11 41 54](https://github.com/GetJobber/atlantis/assets/104797704/596c9681-8fd4-428d-ab8b-478838610773)


### Changed


Update to Chip [here](https://github.com/GetJobber/atlantis/pull/1610/files#diff-af91b758ebbfe83ef19aa7898e882479b307e8178e11e2aa5defaf07fa063240R67); the `chipBar` needs to only be visible when the label is also present

`useComboboxValidation` also had to be updated so that consumers can implement `Combobox` without explicitly setting a `Trigger` (because `ComboboxTrigger` will then be the default).  The verbiage in the errors will change when `ComboboxTriggerChip` & `ComboboxTriggerButton` are officially deleted.

### Deprecated


`ComboboxTriggerButton` and `ComboboxTriggerChip` have been deprecated.

Once this PR is merged, we can double check usages of `Combobox` outside of Atlantis and then **delete**
 these two subcomponents as they are no longer needed.
 

## Testing


Use Storybook to test that `ComboboxTrigger` works as expected:

In the Basic (single select) example:
- select an option and make sure the Chip label changes to that option
- click on the Chip again to reopen the `Combobox` and make a new selection, making sure that new option is now the label of the Chip

In the Multi Select example:
- as you make selections, the options should be immediately load into the Chip label 
- when there is more than one selection, the Chip label should be separating options with a comma
- clicking outside of the `Combobox` or on the trigger will close the element and the Chip still reads `Heading | Option 1, Option 2`


---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)
